{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5b5ee5c4",
   "metadata": {},
   "source": [
    "This notebook uploads two outputs to Google Cloud Storage buckets.\n",
    "\n",
    "1) The inputs array converted to a vcf-style hail table.\n",
    "2) The next upload is a JSON file of the input variants converted to GA4GH GKS VA/VRS annotations using gnomAD data.\n",
    "\n",
    "It can read from a locally-staged hail table directory, or use one in GCS. The table must have a pre-computed `.info.VRS` field. This is computed using `tgg_methods` `vrs_annotation_batch.py`, which uses the `vrs-python` `vcf_annotation.py`\n",
    "\n",
    "https://github.com/broadinstitute/tgg_methods/blob/master/vrs/vrs_annotation_batch.py (last checked at `a0002f02fbd5dd25487b261e94081a3daec29c64`)\n",
    "\n",
    "This validates gnomad_methods functions `gnomad_gks` and `get_gks`, off the branch in this pull request: https://github.com/broadinstitute/gnomad_methods/pull/556\n",
    "\n",
    "The JSON schema is off the `aw-lb-drafting` branch (at `bf3b5aa`)\n",
    "\n",
    "The cell that patches the gnomad_gks function can be removed if the python environment this notebook is in is run off the version of code in the gnomad_methods PR 556."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc240ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e4985d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The clingen-public bucket will only be the week of 2023-09-18\n",
    "# The clingen-public-requesterpays bucket will remain available but the client will need to \n",
    "# be authenticated with a Google Cloud account with billing enabled to pay network transfer fees\n",
    "\n",
    "# configuration for data outputs\n",
    "bucket = \"clingen-public-requesterpays\"\n",
    "bucket = \"clingen-public\"\n",
    "\n",
    "# Writes inputs array as a hail table to this destination, if not None.\n",
    "# This can be useful for other testing, using this hail table as input without reconstructing it\n",
    "# inputs_ht_destination_url = f\"gs://{bucket}/gnomad-gks-qc/gnomad-filtered-inputs.ht\"\n",
    "inputs_ht_destination_url = None\n",
    "\n",
    "# Copies output annotations as newline delimited json to this url, if not None\n",
    "outputs_destination_file = f\"gs://{bucket}/gnomad-gks-qc/outputs.ndjson\"\n",
    "\n",
    "\n",
    "# ht_url can be a gs:// path, or a file:// local path\n",
    "\n",
    "# Publicly readable, but doesn't have all gnomad variants in it\n",
    "# ht_url = f\"gs://{bucket}/gnomad-gks-downsampled-100k.ht\"\n",
    "\n",
    "# Can refer to a local hail table directory\n",
    "ht_url = \"../downsample_to_100k_full_release.ht\"\n",
    "\n",
    "# gnomad-vrs-io-finals is a private bucket, needs a service account with read access \n",
    "# ht_url = \"gs://gnomad-vrs-io-finals/ht-outputs/release_0426_dmtest_v3.1.2-Full-ht-release-output-updated-schema-050523.ht\"\n",
    "\n",
    "\n",
    "print(ht_url)\n",
    "ht = hl.read_table(ht_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edac5b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gnomad\n",
    "import gnomad.utils.annotations\n",
    "import gnomad.resources.grch38.gnomad\n",
    "\n",
    "# reload (re-running this cell will reload modifications to these modules on disk)\n",
    "import importlib\n",
    "importlib.reload(gnomad.utils.annotations)\n",
    "importlib.reload(gnomad.resources.grch38.gnomad)\n",
    "\n",
    "from gnomad.utils.annotations import add_gks_va, add_gks_vrs\n",
    "from gnomad.resources.grch38.gnomad import gnomad_gks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5d3a56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import shutil\n",
    "va_spec_clone = \"gnomad-gks-v1_va-spec\"\n",
    "va_spec_branch = \"gk-pilot\"\n",
    "shutil.rmtree(va_spec_clone, ignore_errors=True)\n",
    "p = subprocess.run([\"git\", \"clone\", \"https://github.com/ga4gh/va-spec\", \"gnomad-gks-v1_va-spec\"],\n",
    "                   check=True)\n",
    "p = subprocess.run([\"bash\", \"-c\",\n",
    "                    f\"cd {va_spec_clone} && git checkout {va_spec_branch}\"],\n",
    "                   check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf2e0aa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open(f\"{va_spec_clone}/schema/cohortAlleleFreq.json\") as f:\n",
    "    schema = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94145a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GnomAD 3.1.2 \n",
    "# GRCh38 expressions\n",
    "\n",
    "inputs = [\n",
    "    {\n",
    "        \"clinvar\": \"422227\",\n",
    "        \"gnomad\": \"16-68737366-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"619\",\n",
    "        \"gnomad\": \"12-102843683-C-T\"\n",
    "    },\n",
    "    {\n",
    "       \"clinvar\": \"808527\",\n",
    "       \"gnomad\": \"19-38442434-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"695823\",\n",
    "        \"gnomad\": \"1-11130641-G-A\"\n",
    "    },\n",
    "    # Not in gnomAD 3.1.2\n",
    "#     {\n",
    "#         \"clinvar\": \"251162\",\n",
    "#         \"gnomad\": \"19-11105249-C-T\"\n",
    "#     },\n",
    "    {\n",
    "        \"clinvar\": \"143344\",\n",
    "        \"gnomad\": \"X-154030690-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"137390\",\n",
    "        \"gnomad\": \"14-28767768-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"129997\",\n",
    "        \"gnomad\": \"15-89317458-C-G\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"89198\",\n",
    "        \"gnomad\": \"2-47799491-C-G\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"43032\",\n",
    "        \"gnomad\": \"14-23416241-G-A\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"2356\",\n",
    "        \"gnomad\": \"1-216247118-C-A\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"3038\",\n",
    "        \"gnomad\": \"11-108248927-T-G\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"256155\",\n",
    "        \"gnomad\": \"7-107701083-T-A\"\n",
    "    },\n",
    "    # example variant from va-spec simple_result_example.yaml\n",
    "    {\n",
    "        \"gnomad\": \"1-55051215-G-GA\"\n",
    "    }\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35b7333",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell constructs an indexed hail table from the input alleles\n",
    "# so the full gnomad dataset can be rapidly filtered by left joining it onto this.\n",
    "# Using a .filter() to check membership in the input set is slow because it needs\n",
    "# to do a table scan.\n",
    "# Too slow:\n",
    "# ht_with_coords.filter(input_gnomad_expressions.contains(ht_with_coords.genomic_coordinates))\n",
    "\n",
    "import pandas\n",
    "\n",
    "input_gnomad_expressions = hl.literal([x[\"gnomad\"] for x in inputs])\n",
    "input_terms = [x[\"gnomad\"].split(\"-\") for x in inputs]\n",
    "\n",
    "df = pandas.DataFrame(\n",
    "    {\n",
    "        \"contig\": [str(\"chr\" + i[0]) for i in input_terms],\n",
    "        \"position\": [int(i[1]) for i in input_terms],\n",
    "        \"ref\": [i[2] for i in input_terms],\n",
    "        \"alt\": [i[3] for i in input_terms]\n",
    "    }\n",
    ")\n",
    "input_ht = hl.Table.from_pandas(df)\n",
    "input_ht = (input_ht\n",
    "    .annotate(\n",
    "        locus=hl.locus(input_ht.contig, input_ht.position, reference_genome=\"GRCh38\"),\n",
    "        alleles=hl.array([input_ht.ref, input_ht.alt]))\n",
    "    .drop(\"contig\", \"position\", \"ref\", \"alt\")\n",
    "    .key_by(\"locus\", \"alleles\"))\n",
    "\n",
    "input_ht.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f65c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "ht_with_coords = ht.annotate(\n",
    "    genomic_coordinates = hl.format(\"%s-%s-%s-%s\",\n",
    "        ht.locus.contig[3:], # Remove 'chr'\n",
    "        hl.str(ht.locus.position),\n",
    "        ht.alleles[0],\n",
    "        ht.alleles[1]\n",
    "    )\n",
    ")\n",
    "ht_with_coords.show()\n",
    "\n",
    "ht_filtered = ht_with_coords\n",
    "\n",
    "# Indexed filter using join on the indexed vcf table constructed from inputs\n",
    "# To enable filtering to the set of input variants, uncomment everything below\n",
    "\n",
    "# ht_filtered = input_ht.join(ht_with_coords)\n",
    "# ht_filtered.genomic_coordinates.show()\n",
    "\n",
    "# # Write the filtered gnomad table to storage\n",
    "# if inputs_ht_destination_url:\n",
    "#     ht_filtered.write(inputs_ht_destination_url, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a6b9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for gnomad_gks/get_gks\n",
    "ancestry_group_short_names = gnomad.resources.grch38.gnomad.POPS[\"v3\"]\n",
    "ancestry_groups_full_name_map = gnomad.sample_qc.ancestry.POP_NAMES\n",
    "gnomad_version_label = \"3.1.4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d6092b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "vrs_variants = []\n",
    "records = ht_filtered.select(\n",
    "    ht_filtered.freq,\n",
    "    ht_filtered.popmax,\n",
    "    ht_filtered.info,\n",
    "    ht_filtered.genomic_coordinates,\n",
    "    vrs=ht_filtered.info.vrs\n",
    ")\n",
    "records = records.take(5)\n",
    "\n",
    "variant_strs = [r.genomic_coordinates for r in records]\n",
    "loci = [\n",
    "    hl.locus(contig=str(\"chr\" + v0), pos=int(v1), reference_genome=\"GRCh38\")\n",
    "    for [v0, v1, *_] in \n",
    "    [v.split(\"-\") for v in variant_strs]\n",
    "]\n",
    "\n",
    "print(f\"gnomAD allele strings: {variant_strs}\")\n",
    "for record in records:\n",
    "    print(\"calling add_gks_vrs on: \" + record.genomic_coordinates)\n",
    "    vrs_variant = add_gks_vrs(\n",
    "        input_vrs=record.info.vrs,\n",
    "        input_locus=record.locus\n",
    "    )\n",
    "    print(json.dumps(vrs_variant, indent=2))\n",
    "    vrs_variants.append(vrs_variant)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f66c7db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell takes one locus present in the input table, performs VRS and VA annotation and prints it.\n",
    "\n",
    "# For performance on larger datasets, the interval should be much larger, at least a few thousand.\n",
    "\n",
    "ivl_0 =hl.locus_interval(loci[0].contig, loci[0].position, loci[0].position+1, reference_genome=\"GRCh38\")\n",
    "ht_locus_0 = (hl.filter_intervals(ht_filtered, [ivl_0]))\n",
    "\n",
    "import time\n",
    "t0 = time.time()\n",
    "gks_annotation = gnomad_gks(\n",
    "    locus_interval=ivl_0,\n",
    "    custom_ht=ht_filtered,\n",
    "    version=\"3.1.4\",\n",
    "    data_type=\"genomes\",\n",
    "    by_ancestry_group=False,\n",
    "    by_sex=False,\n",
    "    vrs_only=False,\n",
    "    skip_checkpoint=True,\n",
    "    skip_coverage=True\n",
    ")\n",
    "t1 = time.time()\n",
    "print(f\"{t1-t0} seconds\")\n",
    "\n",
    "\n",
    "gks_annotation = gnomad_gks(\n",
    "    locus_interval=ivl_0,\n",
    "    custom_ht=ht_filtered,\n",
    "    version=\"3.1.4\",\n",
    "    data_type=\"genomes\",\n",
    "    by_ancestry_group=False,\n",
    "    by_sex=False,\n",
    "    vrs_only=False,\n",
    "    skip_checkpoint=True,\n",
    "    skip_coverage=False\n",
    ")\n",
    "t1 = time.time()\n",
    "print(f\"{t1-t0} seconds\")\n",
    "\n",
    "\n",
    "print(json.dumps(gks_annotation[0], indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cc1fe66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell takes one locus present in the input table, performs VRS and VA annotation and prints it.\n",
    "\n",
    "# For performance on larger datasets, the interval should be much larger, at least a few thousand.\n",
    "print(hl.eval(loci[0].contig))\n",
    "ivl_0 =hl.locus_interval(loci[0].contig, loci[0].position, loci[0].position+1, reference_genome=\"GRCh38\")\n",
    "ht_locus_0 = (hl.filter_intervals(ht_filtered, [ivl_0]))\n",
    "\n",
    "\n",
    "larger_interval = hl.locus_interval(contig=\"chr1\",\n",
    "                       start=1,\n",
    "                       end=int(1e7),\n",
    "                       reference_genome=\"GRCh38\")\n",
    "\n",
    "import time\n",
    "t0 = time.time()\n",
    "gks_annotations = gnomad_gks(\n",
    "    locus_interval=larger_interval,\n",
    "    custom_ht=ht_filtered,\n",
    "    version=\"3.1.4\",\n",
    "    data_type=\"genomes\",\n",
    "    by_ancestry_group=False,\n",
    "    by_sex=False,\n",
    "    vrs_only=False,\n",
    "    skip_checkpoint=True,\n",
    "    skip_coverage=False\n",
    ")\n",
    "t1 = time.time()\n",
    "l = len(gks_annotations)\n",
    "td = t1 - t0\n",
    "print(f\"Annotated {l} records in {td} seconds ({td/l:.2f} sec/rec)\")\n",
    "\n",
    "print(json.dumps(gks_annotation[0], indent=2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
